docs/practices/workflows/profiling.md

```diff
@@ -0,0 +1,103 @@
+```
+
+<blockquote style="background: rgba(255, 200, 0, 0.1); border: 5px solid rgba(255, 200, 0, 0.4);">
+
+Beware that this gives access to certain kernel state and environment to the unprivileged user.
+Once investigation is over either set these properties back to the more restricted settings or,
+better yet, reboot.
+
+Definitely do not run untrusted code after running these commands.
+
+</blockquote>
+
+Then collect a profile as such:
+
+```command
+$ perf record -e cpu-clock -F1000 -g --call-graph dwarf,65528 YOUR_COMMAND_HERE
+```
+
+(you may omit `-e cpu-clock` in certain environments for a more precise sampling timer.)
+
+This will produce a profile file in the current working directory. Although you can inspect the
+profile already with `perf report`, we've had much better experience with using [Firefox
+Profiler](https://profiler.firefox.com/) as the viewer. Although Firefox Profiler supports `perf`
+and many other different data formats, for `perf` in particular a conversion step is necessary:
+
+```command
+$ perf script -F +pid > mylittleprofile.script
+```
+
+Then, load this `mylittleprofile.script` with the tool.
+
+## Memory profiling
+
+`neard` is a pretty memory-intensive application with many allocations occurring constantly.
+Although Rust makes it pretty hard to introduce memory problems, it is still possible to leak
+memory or to inadvertently retain too much of it.
+
+Unfortunately, “just” throwing a random profiler at neard does not work for many reasons. Valgrind
+for example is introducing enough slowdown to significantly alter the behaviour of the run, not to
+mention that to run it successfully and without crashing it will be necessary to comment out
+`neard`’s use of `jemalloc` for yet another substantial slowdown.
+
+So far the only took that worked out well out of the box was
+[`bytehound`](https://github.com/koute/bytehound). Using it is quite straightforward, but needs
+Linux, and ability to execute privileged commands.
+
+First, checkout and build the profiler (you will need to have nodejs `yarn` thing available as
+well):
+
+```command
+$ git clone git@github.com:koute/bytehound.git
+$ cargo build --release -p bytehound-preload
+$ cargo build --release -p bytehound-cli
+```
+
+You will also need a build of your `neard`, once you have that, give it some ambient cabapilities
+necessary for profiling:
+
+```command
+$ sudo setcap 'CAP_SYS_RESOURCE+ep' neard
+```
+
+And finally run the program with the profiler enabled (in this case `neard run` command is used):
+
+```command
+$ env LD_PRELOAD=/path/to/libbytehound.so neard run
+```
+
+### Viewing the profile
+
+<blockquote style="background: rgba(255, 200, 0, 0.1); border: 5px solid rgba(255, 200, 0, 0.4);">
+
+Do note that you will need about twice the amount of RAM as the size of the input file in order to
+load it successfully.
+
+</blockquote>
+
+Once enough profiling data has been gathered, terminate the program. Use the `bytehound` CLI tool
```

Comment:
How are you terminating neard? If I use SIGTERM or SIGINT I get no output file from the profiler. I do see an output `.dat` if I run a command which terminates on its own (e.g. `env LD_PRELOAD=/home/ubuntu/libbytehound.so ls`).

Reply:
Just ^C. For me the profile file gets written continuously throughout the execution…

Reply:
I think the reason why it didn't work for you were incomplete instructions and the fact that the libbytehound was not getting attached... let me update them instructions.